src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs

```diff
@@ -147,5 +149,114 @@ void Initialise()
+                    timer.Change(TimeSpan.FromMinutes(2), Timeout.InfiniteTimeSpan);
+                }
+                catch (ObjectDisposedException)
+                {
+                }
+            }
+
+            void Sync()
+            {
+                using (var session = store.OpenSession())
+                {
+                    var storedHeartbeats = session.Query<Heartbeat>().ToDictionary(h=> h.Id);
+                    var inmemoryHeartbeats = HeartbeatStatusProvider.HeartbeatsPerInstance;
+
+                    foreach (var memHeartbeat in inmemoryHeartbeats)
+                    {
+                        Heartbeat heartbeat;
+                        if (storedHeartbeats.TryGetValue(memHeartbeat.Key, out heartbeat))
+                        {
+                            if (heartbeat.Disabled)
+                            {
+                                continue;
+                            }
+
+                            heartbeat.LastReportAt = memHeartbeat.Value.ExecutedAt;
+                            heartbeat.EndpointDetails = new EndpointDetails
+                            {
+                                HostId = memHeartbeat.Value.HostId,
+                                Host = memHeartbeat.Value.Host,
+                                Name = memHeartbeat.Value.EndpointName
+                            };
+
+                            if (heartbeat.ReportedStatus == Status.Dead)
+                            {
+                                heartbeat.ReportedStatus = Status.Beating;
+                                Bus.Publish(new EndpointHeartbeatRestored
+                                {
+                                    Endpoint = heartbeat.EndpointDetails,
+                                    RestoredAt = heartbeat.LastReportAt
+                                });
+                            }
+                        }
+                        else
+                        {
+                            heartbeat = new Heartbeat
+                            {
+                                Id = memHeartbeat.Key,
+                                ReportedStatus = Status.Beating,
+                                LastReportAt = memHeartbeat.Value.ExecutedAt,
+                                EndpointDetails = new EndpointDetails
+                                {
+                                    HostId = memHeartbeat.Value.HostId,
+                                    Host = memHeartbeat.Value.Host,
+                                    Name = memHeartbeat.Value.EndpointName
+                                }
+                            };
+
+                            Bus.Publish(new HeartbeatingEndpointDetected
+                            {
+                                Endpoint = heartbeat.EndpointDetails,
+                                DetectedAt = heartbeat.LastReportAt,
+                            });
+                        }
+
+                        session.Store(heartbeat);
```

Comment:
Should only need to do this for new heartbeats. If this is an existing heartbeat, it is already tracked as a part of the session.